ohos: fix ohos.c test case error with LLVM_ENABLE_PER_TARGET_RUNTIME_…

phuang opened the pull request:
…DIR=OFF

The problem is because libclang_rt.builtins-{arch}.a for all the arches will be installed into the same folder with old compiler rt layout (LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF) and OHOS driver will get a wrong library in this case. To fix the problem, `-ohos` suffix will be added to the ohos builtin libraries search path.

Note: OHOS driver doesn't support the old layout, compiler-rt for ${arch}-linux-unknown-ohos targets have to be built with
LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=ON

phuang committed:
ohos: fix ohos.c test case error with LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF

The problem is because libclang_rt.builtins-{arch}.a for all the
arches will be installed into the same folder with old compiler rt
layout (LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF) and OHOS driver will
get a wrong library in this case. To fix the problem, `-ohos`
suffix will be added to the ohos builtin libraries search path.

Note: OHOS driver doesn't support the old layout,
${arch}-linux-unknown-ohos targets have to be built with
LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=ON

phuang commented:
> > Note: OHOS driver doesn't support the old layout, compiler-rt for ${arch}-linux-unknown-ohos targets have to be built with LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=ON
> 
> Can we make `cmake` error out if you try to build for `*-linux-unknown-ohos` with `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF`?

Sure. I will try adding.
And one question: will llvm deprecate the old layout? should we consider adding the old layout support for ohos?

carlocab commented:
> And one question: will llvm deprecate the old layout?

There was some desire to, but it seems to have fizzled out: https://discourse.llvm.org/t/rfc-time-to-drop-legacy-runtime-paths/64628

I imagine it'll happen eventually, but it will probably take a while.

> should we consider adding the old layout support for ohos?

I don't know enough about OHOS and its users to be able to say. If you're not sure: I suggest leaving out support for now, and someone who needs it can contribute support for it. But if we do leave support out: good error messages for users who happen to try to use/build it in the wrong configuration are probably essential.

phuang commented:
To add this cmake configure error, where should I look at? any hint?

phuang committed:
ohos: make configure fail if LLVM_ENABLE_PER_TARGET_RUNTIME_DIR is OFF

OHOS driver doesn't support old runtime libraries layout, so make
cmake configure report error if LLVM_ENABLE_PER_TARGET_RUNTIME_DIR
is OFF.

phuang commented:
Updated the cmake script to output error message if LLVM_ENABLE_PER_TARGET_RUNTIME_DIR is OFF for ohos targets.

MaskRay commented:
We can disable the test on -DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF layouts, but I don't find a lit feature that.

The usual way is to create a compiler-rt file hierarchy under `Inputs`. Just `fd libclang_rt.builtins.a` for some examples.

MaskRay reviewed (changes requested):
.
